### PR TITLE
POLIO-758 / POLIO-766: Implement  budget override

### PR DESCRIPTION
**Pull request by olethanh**
Explain what problem this PR is resolving

## Self proof reading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] Did I add translations
- [x] My migrations file are included
- [NO] Are there enough tests
## Changes

Add /override/ endpoint on budgetstep. see main ticket for spec
Add from_node and to_node on BudgetStep and migration to fill them
Add mitigation code for unkown transitions

## How to test

See budget setup on [wiki](https://wiki.bluesquare.org/en/Dev-team/product-management/iaso/polio_budget_workflow)
use the browsable api page here to add a new step http://localhost:8081/api/polio/budget/override/

### CODE REVIEW

**Comment by olethanh** [comment]
@beygorghor  Merged the front PR so the api.ts is easier to review. see the front PR for instruction on how to test https://github.com/BLSQ/iaso/pull/269

File: plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx [hunk 1]
```diff
@@ -13,72 +18,102 @@ import {
...
 }) => {
     const currentUser = useCurrentUser();
     const { data: userHasTeam } = useUserHasTeam(currentUser?.user_id);
     const { formatMessage } = useSafeIntl();
-    const { mutateAsync: saveBudgetStep } = useSaveBudgetStep();
-    const { data: teamOptions, isFetching: isFetchingTeams } =
-        useGetTeamsDropDown();
-    // TODO replace with endpoint when available
-    const stepOptions = [];
+    // TODO replace
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx]: replace what ?

File: plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx [hunk 2]
```diff
@@ -13,72 +18,102 @@ import {
...
-            files: null,
-            links: null,
-            amount: null,
-            target_teams: [],
-            transition: null,
+            comment: undefined,
+            files: undefined,
+            links: undefined,
+            amount: undefined,
+            // this value is for handling non-field arrors from api
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx]: `errors `

File: plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx [hunk 3]
```diff
@@ -92,29 +127,53 @@ const CreateOverrideStep: FunctionComponent<Props> = ({
...
+            if ((links?.length ?? []) > 0) {
+                setFieldTouched('links', true);
+            }
+            if (amount) {
+                setFieldTouched('amount', true);
+            }
+        }
+    }, [setFieldTouched, touched.amount, touched.links, values]);
 
+    const allowConfirm = isValid;
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx]: a bit overkill no ?

File: plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx [hunk 4]
```diff
@@ -130,100 +189,140 @@ const CreateOverrideStep: FunctionComponent<Props> = ({
...
-                                onChange(
-                                    keyValue,
-                                    commaSeparatedIdsToArray(value),
-                                );
+                    <Box mt={2}>
+                        <FilesUpload
+                            files={values.files ?? []}
+                            onFilesSelect={files => {
+                                setFieldTouched('files', true);
+                                setFieldValue('files', files);
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx:231]: `onChange ` ?

File: plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx [hunk 5]
```diff
@@ -130,100 +189,140 @@ const CreateOverrideStep: FunctionComponent<Props> = ({
...
+                        <AddMultipleLinks
+                            required={requiredFields.includes('links')}
                         />
-                        {/* @ts-ignore */}
-                        {(errors?.general ?? []).length > 0 && (
+                    </Box>
+                    <Box mt={2} mb={2}>
+                        <Divider />
+                    </Box>
+                    {(recipientTeams ?? []).length > 0 && (
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx:248]: `recipientTeams?.length > 0 ` ?
 undefined > 0 = false

File: plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx [hunk 6]
```diff
@@ -130,100 +189,140 @@ const CreateOverrideStep: FunctionComponent<Props> = ({
...
+                                })}
+                            </Box>
+                            <Box mt={2} mb={2}>
+                                <Divider />
+                            </Box>
+                        </>
+                    )}
+
+                    {/* @ts-ignore */}
+                    {(errors?.general ?? []).length > 0 && (
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx:279]: Same here ?

File: plugins/polio/js/src/pages/Budget/hooks/api/useSaveOverrideStep.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    amount?: number;
+    campaign_id: string; // uuid
+};
+
+type PostRequestBody = PostArg;
+
+const postOverrideStep = (body: Payload): Promise<BudgetStep> => {
+    const filteredParams = Object.fromEntries(
+        Object.entries(body).filter(
+            // eslint-disable-next-line no-unused-vars
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Budget/hooks/api/useSaveOverrideStep.ts]: you can remove this i think

**Review by beygorghor — APPROVED** [review]
This is working for me, only small remarks on the code.
Also can we improve this error message ?
Screenshot 2023-01-03 at 10 35 08

**Pull request merged**